### PR TITLE
Quality of Life improvement (Queue Drag)

### PR DESCRIPTION
**Pull request by BaalEvan**
Hi!
So I really like your mod, but changing order and forcing to start research on something without losing a queue was something I missed. So here I am :)

Add research to beginning of queue by Ctrl
Add changing queue order by drag

Info about RealQueueRect:
QueueRect Calculated by SetRects was wrong, wrong size and position no idea why, but when i commented setting it in SetRects there was no queue at all, perhaps i missed something not sure, so i added RealQueueRect which is set by Queue during draw.

Info about Translation:
I added features but I didn't modified translation so there is no indication about ability to use CTRL in tooltips.

### CODE REVIEW

**Comment by FluffierThanThou** [comment]
very interesting, will have a look!

File: Source/Queue/Queue.cs [hunk 1]
```diff
@@ -49,6 +49,53 @@ public static void TryDequeue( ResearchNode node )
...
+            if (node.RealQueueRect.Contains(dropPosition)) return false;
+
+            // Order Queue by distance to mouse 
+            var queueByDistance = _instance._queue.OrderBy(item => Vector2.Distance(item.RealQueueRect.center, dropPosition));
+            var nearestItem = queueByDistance.First();
+            var indexToPlace = _instance._queue.IndexOf(nearestItem);
+
+            // Move node to new Position
+            _instance._queue.Remove(node);
+            _instance._queue.Insert(indexToPlace, node);
```
Comment (FluffierThanThou) [inline Source/Queue/Queue.cs:66]: looks like this will fail if node's `index < indexToPlace`, you'll effectively insert at indexToPlace +1

Comment (BaalEvan) [reply Source/Queue/Queue.cs:66]: I tested it, and didn't get any error in any case.

IndexToPlace is saved before node is removed from list and indexes move by -1.
So if you have 10elements and want to place at last position then insert will act as Add.

example from C# Interactive
```
Loading context from 'CSharpInteractive.rsp'.
Type "#help" for more information.
> var q = new List<int>() {0,1,2,3,4,5,6,7,8,9};
> q.Count
10
> var indexToPlace = q.IndexOf(9);
> var node = q[5];
> q.Remove(node);
> q.Insert(indexToPlace,node);
> q.Count
10
> q[indexToPlace]
5
> foreach(var n in q){Console.WriteLine(n);}
0
1
2
3
4
6
7
8
9
5
```

Comment (BaalEvan) [reply Source/Queue/Queue.cs]: Ugh i replied inside code:

I tested it, and didn't get any error in any case.

IndexToPlace is saved before node is removed from list and indexes move by -1.
So if you have 10elements and want to place at last position then insert will act as Add.

example from C# Interactive

Loading context from 'CSharpInteractive.rsp'.
Type "#help" for more information.
> var q = new List<int>() {0,1,2,3,4,5,6,7,8,9};
> q.Count
10
> var indexToPlace = q.IndexOf(9);
> var node = q[5];
> q.Remove(node);
> q.Insert(indexToPlace,node);
> q.Count
10
> q[indexToPlace]
5
> foreach(var n in q){Console.WriteLine(n);}
0
1
2
3
4
6
7
8
9
5

Comment (FluffierThanThou) [reply Source/Queue/Queue.cs:66]: wow, that was fast.

If you're online now, wanna chat on discord? I'm Fluffy on the unofficial official RimWorld discord: https://discord.gg/rimworld

**Comment by MadaraUchiha** [comment]
Hey! What's the status on this? Did you guys reach a resolution on Discord? 😄, I really like this feature.

**Comment by FluffierThanThou** [comment]
It's going to get implemented in one form or another, but not right now.
I'm in the middle of a larger overhaul, and it doesn't really make sense to
work on stuff that will likely need to be changed again in the overhaul
anyway.

**Comment by Spaniard829** [comment]
@BaalEvan Great work, thank you. No in-game errors noticed.

**Pull request closed without merging**